code/modules/procedural mapping/mapGeneratorModule.dm

```diff
@@ -38,24 +34,61 @@
 
 	//Turfs don't care whether atoms can be placed here
 	for(var/turfPath in spawnableTurfs)
-		if(clusterCheckFlags & CLUSTER_CHECK_TURFS)
-			if(clusterMax && clusterMin)
+
+		//Clustering!
+		if(clusterMax && clusterMin)
+
+			//You're the same as me? I hate you I'm going home
+			if(clusterCheckFlags & CLUSTER_CHECK_SAME_TURFS)
 				clustering = rand(clusterMin,clusterMax)
-				if(locate(/atom/movable) in range(clustering, T))
+				for(var/turf/F in trange(clustering,T))
+					//if(F.type == turfPath) //NOT istype(), exact typematching
+					if(istype(F,turfPath))
+						continue
+
+				if(locate(turfPath) in trange(clustering, T))
 					continue
+
+			//You're DIFFERENT to me? I hate you I'm going home
+			if(clusterCheckFlags & CLUSTER_CHECK_DIFFERENT_TURFS)
+				clustering = rand(clusterMin,clusterMax)
+				for(var/turf/F in trange(clustering,T))
+					//if(F.type != turfPath)
+					if(!(istype(F,turfPath)))
+						continue
+
+		//Success!
 		if(prob(spawnableTurfs[turfPath]))
 			T.ChangeTurf(turfPath)
```

Comment:
This isn't an issue, this is never reached if the previous checks fail, due to intelligent use of continue and break statements.